airbase/download_api/download_session.py

```diff
@@ -0,0 +1,259 @@
+        }
+        if skip_existing:
+            url_paths = {
+                url: path
+                for url, path in url_paths.items()
+                # re-download empty files
+                if not path.is_file() or path.stat().st_size == 0
+            }
+
+        # create missing country sum-directories before downloading
+        for parent in {path.parent for path in url_paths.values()}:
+            parent.mkdir(exist_ok=True)
+
+        async for path in tqdm(
+            self.client.download_binary_files(
+                url_paths, raise_for_status=raise_for_status
+            ),
+            initial=len(urls) - len(url_paths),
+            total=len(urls),
+            leave=True,
+            disable=not progress,
+            desc="download",
+        ):
+            assert path.is_file(), f"missing {path.name}"
+
+
+def pollutant_id_from_url(url: str) -> int:
+    """
+    numeric pollutant id from urls like
+        http://dd.eionet.europa.eu/vocabulary/aq/pollutant/1
+        http://dd.eionet.europa.eu/vocabularyconcept/aq/pollutant/44/view
+    """
+    if url.endswith("view"):
+        return int(url.split("/")[-2])
+    return int(url.split("/")[-1])
+
+
+async def download(
+    dataset: Dataset,
+    root_path: Path,
+    *,
+    countries: set[str],
+    pollutants: set[str] | None = None,
+    cities: set[str] | None = None,
+    summary_only: bool = False,
+    overwrite: bool = False,
+    quiet: bool = True,
+    raise_for_status: bool = False,
+    session: DownloadSession = DownloadSession(),
+):
+    """
+    request file urls by country|[city]/pollutant and download unique files
+    """
+    if cities:  # one request for each city/pollutant
+        info = dataset.by_city(*cities, pollutant=pollutants)
+    else:  # one request for each country/pollutant
+        if not countries:
+            countries = COUNTRY_CODES
+        info = dataset.by_country(*countries, pollutant=pollutants)
+
+    async with session:
+        if summary_only:
+            summary = await session.summary(
+                *info, progress=not quiet, raise_for_status=raise_for_status
+            )
+            print(
+                "found {numberFiles:_} file(s), ~{size:_} MB in total".format_map(
+                    summary
+                )
+            )
+        else:
+            async for urls in tqdm(
+                session.client.download_urls(
+                    unique := set(info),
+                    raise_for_status=raise_for_status,
+                ),
+                total=len(unique),
+                leave=True,
+                disable=quiet,
+                desc="generate",
+            ):
+                await session.download_to_directory(
+                    root_path,
+                    *urls,
+                    skip_existing=not overwrite,
+                    progress=not quiet,
+                    raise_for_status=raise_for_status,
+                )
```

Comment:
So the reason I [implemented this with asyncio tasks](https://github.com/JohnPaton/airbase/blob/da7584a825029735688e73150cb72423adc959eb/airbase/download_api.py#L302L325) in #56 is so that URLs will be queued for downloading as soon as they are available. With this implementation, we first generate the URLs for one `info`, then download all of those, then generate the next set of urls, then download those, etc. If one `/ParquetFiles/urls` call or one parquet file download is slow, the rest will all have to wait on it. By instead only looping over the `/ParquetFiles/urls` results and then creating tasks for the actual binary file download, we ensure that the downloads are started as soon as possible, regardless of if other calls are being slow. I have seen this make a decent difference in testing while working on #56. I'd prefer to go that route, and update the progress bar using a callback, and just await all the tasks in an `as_completed` at the end.

Let me know if that makes sense, else I'm happy to make a branch to show you what I mean.

Reply:
I wanted to split the URLs in batches in order to avoid creating 4000+ awaitables/jobs/tasks in one go. Ill look #56 in more detail.

I wish I could apply the `asyncio.Semaphore` to the creation of the awaitables given to `as_completed`, but that is not possible.

Reply:
Now all URLs jobs are queued while only one set of files is downloaded.
This should limit the number of awaitables while keeping all the URL jobs in place.

Reply:
Is it a problem to have 4000 tasks? It's not like they will all be executing at the same time, the client limits concurrency so mostly they will just wait.

Reply:
They still take memory, the current setup does not run out of URLs to download.
I'll go back to the previous setup if you insist.